pkg/auth/policy.go

```diff
@@ -0,0 +1,268 @@
+package auth
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"golang.org/x/crypto/openpgp"
+
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/util"
+)
+
+// A Policy encapsulates the behavior a p2 node needs to authorize
+// its actions. It is possible for implementations to rely on other
+// services for these behaviors, so these calls may be slow or
+// transiently fail.
+type Policy interface {
+	// Check if a pod is authorized to be installed and run on this
+	// node. This involves checking that the pod has a valid signature
+	// and that the signer is authorized to install/run the pod. If
+	// the action is authorized, `nil` will be returned.
+	AuthorizePod(
+		id string,
+		manifest io.Reader,
+		signature io.Reader,
+		logger logging.Logger,
+	) error
+
+	// Check if a file digest has a valid signature and that the
+	// signer is authorized to certify the digest. The caller must
+	// separately check that the actual files match the digest. If
+	// the action is authorized, `nil` will be returned.
+	CheckDigest(digest, signature io.Reader) error
+
+	// Release any resources held by the policy implementation.
+	Close()
+}
+
+// auth.Error wraps all errors generated by the authorization layer,
+// allowing errors to carry structured data.
+type Error struct {
```

Comment:
The main reason for this error type is so that structured data can be added to the error (which will wind up in a Logrus entry).  As I see it, the choices are:
1. Don't use structured logging
2. Pass a logger along the call stack and log errors the instant they're generated
3. Pass up exceptions with context information, and log them centrally
I went with #3 because I've used this pattern in past projects, and it's worked decently well. As I've seen strategy #2 applied, low-level errors tend to generate many log messages, one for each layer of the stack they're handled at.